### PR TITLE
Speed up unique_in_order for large ndarrays

### PR DESCRIPTION
**Pull request by bmerry**
Instead of comparing each array to a previous array, hash each array and
look up the hash in a dictionary. While cityhash is not strictly
required, it is much, much faster than the default (sha1) if no fast
3rd-party hasher is found, so it's added to the requirements so that
users won't need to know to install it separately for good performance.

This takes at least 30 seconds off the time to open a large file when
using applycal.

### CODE REVIEW

**Comment by bmerry** [comment]
@ludwigschwardt would you be able to test this out on a few observations from your zoo and a few sorts of sensors to make sure it doesn't break anything?

**Comment by bmerry** [comment]
I've marked this don't-merge-yet because if it's approved, we should put cityhash into katsdpdockerbase and add it to the dependents' requirements.txt files before merging.

**Comment by ludwigschwardt** [comment]
Oops, still wanted to test it on a zoo - let me get back to you shortly...

File: katdal/categorical.py [hunk 1]
```diff
@@ -180,11 +181,14 @@ def unique_in_order(elements, return_inverse=False):
         lookup = collections.OrderedDict(zip(elements, len(elements) * [0]))
     except TypeError:
         # Fall back to slower list-based lookup for unhashable object values
```
[the remark this answers is not in the document]
Comment (bmerry) [reply katdal/categorical.py]: Thanks, fixed.

**Comment by bmerry** [comment]
> I've tested this on a typical dataset with cal products by extracting all sensors as categorical:

Does that mean you think it's all still working?

**Comment by ludwigschwardt** [comment]
Well, the dask tokenizer ran on at least ndarrays, numpy scalars, dicts, lists of floats & complex numbers and tuples of floats, so the coverage is pretty wide and I'm happy :-)

**Pull request merged**